### ISSUE TITLE
Indexing with list of booleans differs from numpy

### ISSUE THREAD

**Issue by fmassa** [pytorch/pytorch#6773]
In numpy, indexing with a list of booleans is equivalent to indexing with a boolean array, which means it performs masking.

In PyTorch, the list of booleans is cast to a long tensor.

I believe this discrepancy should be fixed.

```python
a = torch.rand(3, 3)
print(a)

 0.1041  0.6888  0.7988
 0.9398  0.9151  0.7642
 0.5340  0.4715  0.8128
[torch.FloatTensor of size (3,3)]

print(a[[True, False, False]]

 0.9398  0.9151  0.7642
 0.1041  0.6888  0.7988
 0.1041  0.6888  0.7988
[torch.FloatTensor of size (3,3)]

aa = a.numpy()
print(aa[[True, False, False]])
array([[0.10411686, 0.6887991 , 0.7988465 ]], dtype=float32)
```

**Comment by fmassa** [comment]
I believe a fix would be to replace [this line](https://github.com/pytorch/pytorch/blob/master/torch/csrc/autograd/python_variable_indexing.cpp#L104) with a variant that infers the type automatically, [maybe this function?](https://github.com/pytorch/pytorch/blob/749d51414af11da48bdb195375f702822450aded/torch/csrc/utils/tensor_new.cpp#L174-L176)

**Comment by ruochunz** [comment]
Hi, I was thinking I might give this a try.
This issue seems to come up when boolean lists are used, if i use boolean tensors the behaviour seems to match numpy.

```
a = torch.rand(3,3)
print(a)
tensor([[ 0.0072,  0.8932,  0.4831],
        [ 0.4591,  0.7305,  0.3776],
        [ 0.1386,  0.0242,  0.5616]])

b = torch.tensor([True, False, False])
print(a[b])
tensor([[ 0.0072,  0.8932,  0.4831]])
```
I suspect the `applySlicing` function in the [python_variable_indexing.cpp](https://github.com/pytorch/pytorch/blob/master/torch/csrc/autograd/python_variable_indexing.cpp#L135) file is converting the `[True, False, False]` to long tensor under this [else if statement](https://github.com/pytorch/pytorch/blob/master/torch/csrc/autograd/python_variable_indexing.cpp#L180) with `sequenceToVariable`. 

This is because [dispatch_index](https://github.com/pytorch/pytorch/blob/master/torch/csrc/autograd/python_variable_indexing.cpp#L302) looks like it is responsible for performing the "advanced" indexing with output from 'sequenceToVariable', or a tensor that was used as index, and it gives matched numpy behaviour when boolean tensors are used. Therefore I think the problem is with the output from 'sequenceToVariable'

I will try to make a change to [sequenceToVariable](https://github.com/pytorch/pytorch/blob/master/torch/csrc/autograd/python_variable_indexing.cpp#L108) so that it keeps booleans as uint8 instead of casting everything to long.

Essentially it will check if the entire list is boolean and then create a Variable that has dtype=uint8. 
Just wondering if I'm going down the right track, thanks.

**Comment by fmassa** [comment]
Hi @ruochunz,

I think you are on the right track, and the understanding of the issue is right.

I believe an easy way of keeping the entire list to be boolean is to leverage the `type_inference` functionality that we already have (I believe [it is located here, but I might be wrong](https://github.com/pytorch/pytorch/blob/749d51414af11da48bdb195375f702822450aded/torch/csrc/utils/tensor_new.cpp#L105-L148)), and then assert that it should be either `uint8` or `int64`.

**Comment by ruochunz** [comment]
Hi @fmassa ,

Thanks for the pointers, I've had a look at the [tensor_new.cpp](https://github.com/pytorch/pytorch/blob/749d51414af11da48bdb195375f702822450aded/torch/csrc/utils/tensor_new.cpp) and I found that `sequenceToVariable` calls [legacy_new_from_data](https://github.com/pytorch/pytorch/blob/749d51414af11da48bdb195375f702822450aded/torch/csrc/utils/tensor_new.cpp#L207), which then calls [internal_new_from_data](https://github.com/pytorch/pytorch/blob/749d51414af11da48bdb195375f702822450aded/torch/csrc/utils/tensor_new.cpp#L174). 

It turns out `internal_new_from_data` already has the type_interface implemented, but the type_interface flag is set to False when called upon by `legacy_new_from_data`.

So I made `sequenceToVariable` call a new function `legacy_infer_new_from_data` instead, this function would be exactly the same as `internal_new_from_data` except it calls `internal_new_from_data` with the `type_interface` set to True.

```
// legacy_new_from_data function
Tensor legacy_new_from_data(
    const Type& type,
    c10::optional<Device> device,
    PyObject* data) {
  return internal_new_from_data(type, device, data, false, false, false);
}

// proposed new function, notice the 6th input to internal_new_from_data is set to true 
Tensor legacy_infer_new_from_data(
    const Type& type,
    c10::optional<Device> device,
    PyObject* data){
  return internal_new_from_data(type, device, data, false, false, true);
}
```

I was then able to build this and it produces the numpy behaviour:

```
import torch
a = torch.rand(3,3)
print(a)
tensor([[0.3095, 0.0574, 0.9273],
        [0.5856, 0.0060, 0.8336],
        [0.0340, 0.4121, 0.3204]])

print(a[[True, False, False]])
tensor([[0.3095, 0.0574, 0.9273]])

```

Is this a valid way of solving this issue? Should I do a pull request?

**Comment by fmassa** [comment]
Hi,

Can't you just call `infer_scalar_type` in `sequenceToVariable`, and then enforce that it is either a bool or a long type, and pass this type to `legacy_new_from_data`?

**Comment by fmassa** [comment]
Oh, it's because of the Type/ScalarType differences. I'm not sure, maybe @gchanan knows a bit better?

**Comment by ruochunz** [comment]
Hi @fmassa ,

The `infer_scalar_type` isn't in the [header file](https://github.com/pytorch/pytorch/blob/749d51414af11da48bdb195375f702822450aded/torch/csrc/utils/tensor_new.h), so it can't be called by `sequenceToVariable`. And all the functions in this[ header file](https://github.com/pytorch/pytorch/blob/749d51414af11da48bdb195375f702822450aded/torch/csrc/utils/tensor_new.h) are all involved with creating a tensor, so i wasn't sure if it was appropriate to declare `infer_scalar_type` in there as well.

So instead I declared another function which also has the purpose of creating a tensor, except it is able to enforce if the list is Boolean or Long. If it is ok to declare `infer_scalar_type` in the header file, I would be happy to call it in `sequenceToVariable` :) 

I've also managed to fix the new function so it only runs `infer_scalar_type` once, and it was able to pass the [indexing tests](https://github.com/pytorch/pytorch/blob/master/test/test_indexing.py).

```
Tensor legacy_infer_new_from_data(
    const Type& type,
    c10::optional<Device> device,
    PyObject* data){
  ScalarType scalar_type = infer_scalar_type(data);
  if (scalar_type == ScalarType::Byte){
      auto& idx_type = type.toScalarType(scalar_type);
      return internal_new_from_data(idx_type, device, data, false, false, false);
  }else{
      return internal_new_from_data(type, device, data, false, false, false);
  }
}
```
Just wondering if I should go with this or declare `infer_scalar_type` in the header file and then call with `sequenceToVariable`

**Comment by zou3519** [comment]
We should also fix this in the JIT, I think it implements indexing separately and #14932 doesn't touch those bits.

**Comment by fmassa** [comment]
@zou3519 I'm not sure the JIT fully-supports advanced indexing yet, and I think that a few simple cases break, like `a[[1,2,3], :3, [0, 1, 2]]`

**Comment by zou3519** [comment]
I think everything you just typed above is supported.

**Comment by fmassa** [comment]
Even for assignment? I was thinking about https://github.com/pytorch/pytorch/issues/14332

But you are right that the JIT supports more advanced indexing than I was originally thinking, which is great!
